src/Symfony/Component/HttpFoundation/UrlParser/Url.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation\UrlParser;
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+final class Url implements \Stringable
+{
+    public function __construct(
+        public string $scheme,
+        public ?string $user = null,
+        public ?string $pass = null,
+        public ?string $host = null,
+        public ?int $port = null,
+        public ?string $path = null,
+        public ?string $query = null,
+        public ?string $fragment = null
+    ) {
+    }
+
+    public function isAuthenticated(): bool
+    {
+        return null !== $this->user || null !== $this->pass;
+    }
+
+    public function isScheme(string $scheme): bool
+    {
+        return $this->scheme === $scheme;
+    }
+
+    public function __toString(): string
+    {
+        $dsn = $this->scheme.'://';
+
+        if ($this->isAuthenticated()) {
+            $dsn .= rawurlencode($this->user).':'.rawurlencode($this->pass).'@';
```

Comment:
Will throw if one of `this->user` or `this->pass` is null, as [rawurlencode](https://www.php.net/manual/fr/function.rawurlencode.php)() requires a string

Reply:
Good catch 👍